Common/Securities/SecurityCacheProvider.cs

```diff
@@ -0,0 +1,98 @@
+ * limitations under the License.
+ *
+*/
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace QuantConnect.Securities
+{
+    /// <summary>
+    /// A helper class that will provide <see cref="SecurityCache"/> instances
+    /// </summary>
+    /// <remarks>The value of this class and its logic is performance.
+    /// This class allows for two different <see cref="Security"/> to share the same
+    /// data type cache through different instances of <see cref="SecurityCache"/>.
+    /// This is used to directly access custom data types through their underlying</remarks>
+    public class SecurityCacheProvider
+    {
+        private readonly ConcurrentDictionary<SecurityIdentifier, ConcurrentBag<Symbol>> _relatedSymbols;
+        private readonly IDictionary<Symbol, Security> _securityProvider;
+
+        /// <summary>
+        /// Creates a new instance
+        /// </summary>
+        /// <param name="securityProvider">The security provider to use</param>
+        public SecurityCacheProvider(IDictionary<Symbol, Security> securityProvider)
+        {
+            _securityProvider = securityProvider;
+            _relatedSymbols = new ConcurrentDictionary<SecurityIdentifier, ConcurrentBag<Symbol>>();
+        }
+
+        /// <summary>
+        /// Will return the <see cref="SecurityCache"/> instance to use for a give Symbol.
+        /// If the provided Symbol is a custom type which has an underlying we will try to use the
+        /// underlying SecurityCache type cache, if the underlying is not present we will keep track
+        /// of the custom Symbol in case it is added later.
+        /// </summary>
+        /// <returns>The cache instance to use</returns>
+        public SecurityCache GetSecurityCache(Symbol symbol)
+        {
+            var securityCache = new SecurityCache();
+            if (symbol.SecurityType == SecurityType.Base && symbol.HasUnderlying)
+            {
+                Security underlyingSecurity;
+                if(_securityProvider.TryGetValue(symbol.Underlying, out underlyingSecurity))
+                {
+                    // we found the underlying, lets use its data type cache
+                    SecurityCache.ShareTypeCacheInstance(underlyingSecurity.Cache, securityCache);
+                }
+                else
+                {
+                    // we didn't find the underlying, lets keep track of the underlying symbol which might get added in the future.
+                    // else when it is added, we would have to go through existing Securities and find any which use it as underlying
+                    if (_relatedSymbols.ContainsKey(symbol.Underlying.ID))
+                    {
+                        _relatedSymbols[symbol.Underlying.ID].Add(symbol);
+                    }
+                    else
+                    {
+                        _relatedSymbols[symbol.Underlying.ID] = new ConcurrentBag<Symbol> { symbol };
```

Comment:
is there any chance of this line running on multiple threads? If so, we should prefer to use
``` csharp
_relatedSymbols.AddOrUpdate(symbol.Underlying.ID,
    id => new ConcurrentBag<Symbol>{symbol},
    (id, bag) => { bag.Add(symbol); return bag; }
);
````

Reply:
Will do 👍 - it's even cleaner